tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks.XPlat/DownloadFileSB.cs

```diff
@@ -0,0 +1,315 @@
+/// build it to get rid of the issue. 
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using Microsoft.DotNet.Build.Tasks;
+using System;
+using System.IO;
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.Build.Tasks
+{
+    /// <summary>
+    /// Represents a task that can download a file.
+    /// </summary>
+    public sealed class DownloadFileSB : BuildTask, ICancelableTask
+    {
+        private readonly CancellationTokenSource _cancellationTokenSource = new CancellationTokenSource();
+
+        /// <summary>
+        /// Gets or sets an optional filename for the destination file.  By default, the filename is derived from the <see cref="SourceUrl"/> if possible.
+        /// </summary>
+        public ITaskItem DestinationFileName { get; set; }
+
+        /// <summary>
+        /// Gets or sets a <see cref="ITaskItem"/> that specifies the destination folder to download the file to.
+        /// </summary>
+        [Required]
+        public ITaskItem DestinationFolder { get; set; }
+
+        /// <summary>
+        /// Gets or sets a <see cref="ITaskItem"/> that contains details about the downloaded file.
+        /// </summary>
+        [Output]
+        public ITaskItem DownloadedFile { get; set; }
+
+        /// <summary>
+        /// Gets or sets an optional number of times to retry if possible.
+        /// </summary>
+        public int Retries { get; set; }
+
+        /// <summary>
+        /// Gets or sets the number of milliseconds to wait before retrying.
+        /// </summary>
+        public int RetryDelayMilliseconds { get; set; } = 5 * 1000;
+
+        /// <summary>
+        /// Gets or sets an optional value indicating whether or not the download should be skipped if the file is up-to-date.
+        /// </summary>
+        public bool SkipUnchangedFiles { get; set; } = true;
+
+        /// <summary>
+        /// Gets or sets the URL to download.
+        /// </summary>
+        [Required]
+        public string SourceUrl { get; set; }
+
+        /// <summary>
+        /// Gets or sets a <see cref="HttpMessageHandler"/> to use.  This is used by unit tests to mock a connection to a remote server.
+        /// </summary>
+        internal HttpMessageHandler HttpMessageHandler { get; set; }
+
+        /// <inheritdoc cref="ICancelableTask.Cancel"/>
+        public void Cancel()
+        {
+            _cancellationTokenSource.Cancel();
+        }
+
+        public override bool Execute()
+        {
+            return ExecuteAsync().GetAwaiter().GetResult();
+        }
+
+        private async Task<bool> ExecuteAsync()
+        {
+            if (!Uri.TryCreate(SourceUrl, UriKind.Absolute, out Uri uri))
+            {
+                Log.LogErrorWithCodeFromResources("DownloadFileSB.ErrorInvalidUrl", SourceUrl);
```

Comment:
I see an error in CI that I believe is related to these calls:

```
/src/build.proj(152,5): error MSB4018: The "DownloadFileSB" task failed unexpectedly.
/src/build.proj(152,5): error MSB4018: System.InvalidOperationException: The "DownloadFileSB" task has not registered its resources. In order to use the "TaskLoggingHelper.FormatResourceString()" method this task needs to register its resources either during construction, or via the "TaskResources" property.
/src/build.proj(152,5): error MSB4018:    at Microsoft.Build.Shared.ErrorUtilities.ThrowInvalidOperation(String resourceName, Object[] args)
/src/build.proj(152,5): error MSB4018:    at Microsoft.Build.Utilities.TaskLoggingHelper.FormatResourceString(String resourceName, Object[] args)
/src/build.proj(152,5): error MSB4018:    at Microsoft.Build.Utilities.TaskLoggingHelper.LogErrorWithCodeFromResources(String messageResourceName, Object[] messageArgs)
/src/build.proj(152,5): error MSB4018:    at Microsoft.Build.Tasks.DownloadFileSB.ExecuteAsync() in /src/tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks.XPlat/DownloadFileSB.cs:line 126
/src/build.proj(152,5): error MSB4018:    at Microsoft.Build.Tasks.DownloadFileSB.Execute() in /src/tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks.XPlat/DownloadFileSB.cs:line 79
/src/build.proj(152,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
/src/build.proj(152,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask)
```

I think since the source-build tasks DLL doesn't have these resources, that happens.

Replacing them with ordinary log commands like this isn't quite as easy to diagnose but would be a simple fix:

```suggestion
                Log.LogError($"DownloadFileSB.ErrorInvalidUrl {SourceUrl}");
```

An ideal replacement though would be to look up the string resource and put that in as the format string:

https://github.com/microsoft/msbuild/blob/2791d9d93e88325011eb6907579d6fdac0b1b62e/src/Tasks/Resources/Strings.resx#L2737-L2738

```suggestion
                Log.LogError("MSB3921: The specified URL "{0}" is not valid.", SourceUrl);
```

Even more ideal would be some way to keep using `LogErrorWithCodeFromResources` and share these resource strings, or copy them over, but I'm not familiar enough with resources to give a suggestion in that direction.